### PR TITLE
DT-466 Lookup offender and conviction

### PR DESCRIPTION
**Pull request by andymarke**

### CODE REVIEW

File: src/main/java/uk/gov/justice/digital/delius/service/ConvictionService.java [hunk 1]
```diff
@@ -132,6 +146,28 @@ public Conviction addCourtCaseFor(Long offenderId, CourtCase courtCase) {
...
+            case 1:
+                return events.stream().findFirst();
+            default:
+                throw new DuplicateConvictionsForSentenceDateException(events.size());
+        }
+    }
+
+    private boolean didSentenceStartAroundDate(Event event, LocalDate sentenceStartDate) {
+        // typically used to match start dates in NOMIS and Delius which may be out by a few days
+        return DAYS.between(event.getDisposal().getStartDate(), sentenceStartDate) <= 7;
```
Comment (andymarke) [inline src/main/java/uk/gov/justice/digital/delius/service/ConvictionService.java]: Further analysis required to see if 7 days is the correct leniency between NOMIS and Delius data

File: src/main/java/uk/gov/justice/digital/delius/service/CustodyService.java [hunk 2]
```diff
@@ -108,13 +109,42 @@ public Custody updateCustody(final String nomsNumber,
...
+        final var maybeOffender = offenderRepository.findByNomsNumber(nomsNumber);
+        final var telemetryProperties = Map.of("offenderNo", nomsNumber,
+                "bookingNumber", updateCustodyBookingNumber.getBookingNumber(),
+                "sentenceStartDate", updateCustodyBookingNumber.getSentenceStartDate().format(DateTimeFormatter.ISO_DATE));
+
+        return maybeOffender.map(offender -> {
+            try {
+                final var maybeEvent = convictionService.getSingleActiveConvictionIdByOffenderIdAndCloseToSentenceDate(offender.getOffenderId(), updateCustodyBookingNumber.getSentenceStartDate());
+                return maybeEvent.map(event -> {
+                    // update booking Number
```
Comment (andymarke) [inline src/main/java/uk/gov/justice/digital/delius/service/CustodyService.java]: All the TOD bits in comments, so service still does nothing

File: src/test/java/uk/gov/justice/digital/delius/service/ConvictionServiceTest.java [hunk 3]
```diff
@@ -29,7 +30,7 @@
 import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.when;
 
-@RunWith(SpringRunner.class)
+@ExtendWith(SpringExtension.class)
```
Comment (andymarke) [inline src/test/java/uk/gov/justice/digital/delius/service/ConvictionServiceTest.java]: Refactored to JUnit 5 and introduced nested hence the huge change - really just moved tests into appropriate nested tests

File: src/test/java/uk/gov/justice/digital/delius/service/CustodyServiceTest.java [hunk 4]
```diff
@@ -1,12 +1,14 @@
 package uk.gov.justice.digital.delius.service;
 
 import com.microsoft.applicationinsights.TelemetryClient;
-import org.junit.Before;
-import org.junit.Test;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Test;
```
Comment (andymarke) [inline src/test/java/uk/gov/justice/digital/delius/service/CustodyServiceTest.java]: Also refactored to JUnit 5 and introduced nested hence the huge change - really just moved tests into appropriate nested tests, and added new tests for this change

File: src/main/java/uk/gov/justice/digital/delius/service/CustodyService.java (hunk 2 above)
Comment (matt-crowson-moj) [reply src/main/java/uk/gov/justice/digital/delius/service/CustodyService.java]: Perhaps include TODO in the comment?  Just thinking if you want to search for outstanding stuff.  IDE's can then help out hear searching or even grepping for example....just a thought.

File: src/main/java/uk/gov/justice/digital/delius/service/ConvictionService.java (hunk 1 above)
Comment (mikehalmamoj) [reply src/main/java/uk/gov/justice/digital/delius/service/ConvictionService.java]: Should 7 be a constant?

File: src/main/java/uk/gov/justice/digital/delius/service/CustodyService.java [hunk 5]
```diff
@@ -108,13 +109,43 @@ public Custody updateCustody(final String nomsNumber,
...
-                .build();
+        final var maybeOffender = offenderRepository.findByNomsNumber(nomsNumber);
+        final var telemetryProperties = Map.of("offenderNo", nomsNumber,
+                "bookingNumber", updateCustodyBookingNumber.getBookingNumber(),
+                "sentenceStartDate", updateCustodyBookingNumber.getSentenceStartDate().format(DateTimeFormatter.ISO_DATE));
+
+        return maybeOffender.map(offender -> {
+            try {
+                final var maybeEvent = convictionService.getSingleActiveConvictionIdByOffenderIdAndCloseToSentenceDate(offender.getOffenderId(), updateCustodyBookingNumber.getSentenceStartDate());
+                return maybeEvent.map(event -> {
```
Comment (andrewrlee) [inline src/main/java/uk/gov/justice/digital/delius/service/CustodyService.java]: could do the `orElseThrow` where variable declared - then you wouldn't have to do the `map` and so could avoid large amounts of nesting?

File: src/main/java/uk/gov/justice/digital/delius/service/ConvictionService.java [hunk 6]
```diff
@@ -51,6 +53,18 @@ public int getConvictionCount() {
             return convictionCount;
         }
     }
+    public static class DuplicateConvictionsForSentenceDateException extends Exception {
```
Comment (matt-crowson-moj) [inline src/main/java/uk/gov/justice/digital/delius/service/ConvictionService.java]: This is for my benefit really.

Lots of checked exceptions for controlling flow?  Just seems a bit old school.  Is it historic?  Sorry don't know what is new versus existing...

File: src/main/java/uk/gov/justice/digital/delius/service/CustodyService.java (hunk 5 above)
Comment (andymarke) [reply src/main/java/uk/gov/justice/digital/delius/service/CustodyService.java]: Not sure, the problem is essentially we have a few Optionals, but for each one that we done have a value I want to do something slightly different with the Telemetry event.

`maybeA.elseThrow({do something).somethingElse().elseThrow({do something else})` but I will give it a go next revision.

Luckily all these Telemetry events will be removed once I've done and rolled out, so it should become a flat pipeline.

Thanks

File: src/main/java/uk/gov/justice/digital/delius/service/ConvictionService.java (hunk 6 above)
Comment (andymarke) [reply src/main/java/uk/gov/justice/digital/delius/service/ConvictionService.java]: Good point; we have 3 states; 

1. I have found the thing
2. I have not found the thing
3. I have found loads of things when there should only be 1.

Not sure the best way to handle that other than am Optional and exception.

I suppose I could have a `Result` type class instead. Any better ideas @matt-crowson-moj

File: src/main/java/uk/gov/justice/digital/delius/service/CustodyService.java (hunk 2 above)
Comment (andymarke) [reply src/main/java/uk/gov/justice/digital/delius/service/CustodyService.java]: Good, point, although I'm doing the TODOs as we speak

File: src/main/java/uk/gov/justice/digital/delius/service/ConvictionService.java (hunk 1 above)
Comment (andymarke) [reply src/main/java/uk/gov/justice/digital/delius/service/ConvictionService.java]: Good point, I'll do that next PR

File: src/main/java/uk/gov/justice/digital/delius/service/ConvictionService.java (hunk 6 above)
Comment (matt-crowson-moj) [reply src/main/java/uk/gov/justice/digital/delius/service/ConvictionService.java]: If only java had such a class (Kotlin caters this one).  Of course there are 3rd party libs that support this sort of thing providing Try and Result objects...

**Pull request merged**